rotkehlchen/globaldb/handler.py

```diff
@@ -250,24 +250,28 @@ def retrieve_assets(filter_query: 'AssetsFilterQuery') -> Tuple[List[Dict[str, A
         underlying_tokens: Dict[str, List[Dict[str, str]]] = defaultdict(list)
         prepared_filter_query, bindings = filter_query.prepare()
         parent_query = """
-        SELECT A.identifier AS identifier, A.type, B.address, B.decimals, A.name, C.symbol, C.started, C.forked, C.swapped_for, C.coingecko, C.cryptocompare, B.protocol, B.chain, B.token_kind, D.notes, D.type AS custom_asset_type
-        FROM assets as A
+        SELECT A.identifier AS identifier, A.type, B.address, B.decimals, A.name, C.symbol,
+        C.started, C.forked, C.swapped_for, C.coingecko, C.cryptocompare, B.protocol, B.chain,
+        B.token_kind, D.notes, D.type AS custom_asset_type FROM
+        assets as A
         LEFT JOIN common_asset_details AS C ON C.identifier = A.identifier
         LEFT JOIN evm_tokens as B ON B.identifier = A.identifier
         LEFT JOIN custom_assets as D ON D.identifier = A.identifier
-        """  # noqa: E501
+        """
         query = f'SELECT * FROM ({parent_query}) {prepared_filter_query}'
-        underlying_tokens_query = """
-        SELECT assets.identifier AS identifier, address, token_kind, parent_token_entry, weight, name, symbol, type, chain
-        FROM underlying_tokens_list
-        INNER JOIN assets ON assets.identifier = parent_token_entry
-        INNER JOIN evm_tokens ON assets.identifier=evm_tokens.identifier
-        INNER JOIN common_asset_details ON common_asset_details.identifier = assets.identifier
-        """  # noqa: E501
+        # Get the identifier of the EVM tokens in the filter and query the underlying tokens where
+        # parent_token_entry is in that set of identifiers. We need to join with evm_tokens since
+        # the address column is present there.
+        underlying_tokens_query = (
+            f'SELECT parent_token_entry, address, token_kind, weight FROM underlying_tokens_list '
+            f'LEFT JOIN evm_tokens ON underlying_tokens_list.identifier=evm_tokens.identifier '
+            f'WHERE parent_token_entry IN (SELECT identifier FROM ({query}))'
```

Comment:
Subqueries are generally a non-performant way to retrieve something when compared to a JOIN. And here your subquery is huge. It's ALL of our assets. I think this should be turned to a join. Would also be fun to compare this query versus a join version instead of the subquery.

Depends on the DB optimizer and the particular query, but wonder how much of a diff it would be.

Reply:
This approach is my fault since I was thinking only from the perspective of the application as the consumer. 

When the application is the consumer, this subquery is never all the assets because of pagination. The pagination can be at most 100 items and not more. 

FYI In the case of the 10-100 items per page, the API response time is around 40-90ms